api/provider/aws/test_aws/task_delete_test.go

```diff
@@ -0,0 +1,94 @@
+package test_aws
+
+import (
+	"testing"
+
+	"github.com/aws/aws-sdk-go/service/ecs"
+	"github.com/golang/mock/gomock"
+	provider "github.com/quintilesims/layer0/api/provider/aws"
+	"github.com/quintilesims/layer0/api/tag"
+	awsc "github.com/quintilesims/layer0/common/aws"
+	"github.com/quintilesims/layer0/common/config/mock_config"
+	"github.com/quintilesims/layer0/common/models"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestTaskDelete(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	mockAWS := awsc.NewMockClient(ctrl)
+	tagStore := tag.NewMemoryStore()
+	mockConfig := mock_config.NewMockAPIConfig(ctrl)
+
+	mockConfig.EXPECT().Instance().Return("test").AnyTimes()
+
+	tags := models.Tags{
+		{
+			EntityID:   "tsk_id",
+			EntityType: "task",
+			Key:        "name",
+			Value:      "tsk_name",
+		},
+		{
+			EntityID:   "tsk_id",
+			EntityType: "task",
+			Key:        "arn",
+			Value:      "arn:aws:ecs:region:012345678910:task/arn",
+		},
+		{
+			EntityID:   "tsk_id",
+			EntityType: "task",
+			Key:        "environment_id",
+			Value:      "env_id",
+		},
+	}
+
+	for _, tag := range tags {
+		if err := tagStore.Insert(tag); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	stopTaskInput := &ecs.StopTaskInput{}
+	stopTaskInput.SetCluster("l0-test-env_id")
+	stopTaskInput.SetTask("arn:aws:ecs:region:012345678910:task/arn")
+
+	mockAWS.ECS.EXPECT().
+		StopTask(stopTaskInput).
+		Return(nil, nil)
+
+	target := provider.NewTaskProvider(mockAWS.Client(), tagStore, mockConfig)
+	err := target.Delete("tsk_id")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	assert.Len(t, tagStore.Tags(), 0)
+}
+
+func TestDeleteTaskIdempotence(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	mockAWS := awsc.NewMockClient(ctrl)
+	tagStore := tag.NewMemoryStore()
+	mockConfig := mock_config.NewMockAPIConfig(ctrl)
+
+	// todo: setup helper for config
+	mockConfig.EXPECT().Instance().Return("test").AnyTimes()
+
+	stopTaskInput := &ecs.StopTaskInput{}
+	stopTaskInput.SetCluster("l0-test-env_id")
+	stopTaskInput.SetTask("arn:aws:ecs:region:012345678910:task/arn1")
+
+	mockAWS.ECS.EXPECT().
+		StopTask(stopTaskInput).
+		Return(nil, nil)
+
+	target := provider.NewTaskProvider(mockAWS.Client(), tagStore, mockConfig)
+	err := target.Delete("tsk_id")
+	if err != nil {
+		// t.Fatal(err)
```

Comment:
This also needs the `if err := ...; err != nil { ... }` pattern. Also, it seems that this should be uncommented.

Reply:
This is something I wanted to ask about if you look at the task delete it reads the aws error message. I’m wondering if I should do the same or if we should change both. 
@zpatrick

Reply:
Ah, I see now. So, yes, since this test is testing for idempotence, we're mainly trying to test the graceful handling of an error returned from the SDK. For instance, the SDK helper function [deleteLoadBalancer](https://github.com/quintilesims/layer0/blob/api-refactor/api/provider/aws/load_balancer_delete.go#L54) checks for the presence of a "NoSuchEntity" error code from the SDK and exits gracefully. Likewise [my EXPECT() call](https://github.com/quintilesims/layer0/pull/410/files#diff-ed65c448f68e560743c6c2a4999f9b22R93) ensures that this is happened.

The other helper functions that are called have similar error code checks. In your case, the idempotent check [looks for a specific string in the error message](https://github.com/quintilesims/layer0/blob/api-refactor/api/provider/aws/task_delete.go#L47). @zpatrick has a good example of how to handle this [here](https://github.com/quintilesims/layer0/blob/api-refactor/api/provider/aws/test_aws/environment_delete_test.go#L101)